### PR TITLE
[JAX] Align fused attn backward output gradient sharding and other test strengthening

### PR DESCRIPTION
**Pull request by KshitijLakhani**
# Description

Fix JAX fused-attention bwd sharding and THD metadata handling, while strengthening bwd test coverage with nonuniform output gradients. 

The starting point was fixing the JAX fused-attention bwd sharding, but along the way as I decided to strengthen the testing as part of root causing and then just fixed any ripple effects it had in exposing smaller issues in our feature code

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

1. _Align bwd doutput sharding with the saved fwd output sharding :_ JAX can infer a doutput sharding that is incompatible with the cuDNN bwd kernel, particularly when CP reorder operations are differentiated through. TE now explicitly aligns doutput sharding with the saved output sharding. 

2. _Use nonuniform output gradients in CP tests:_ Existing tests generated a uniform output gradient from mean(output). Uniform values are unchanged by permutation and therefore masked incorrect CP placement. 
NOTE: This was the reason why TE tests did not catch the doutput sharding bug from above and so I decided to go with a more generic case.

3. Route rotated THD ring metadata through the general path: The THD fast metadata path assumes Q and KV describe matching token metadata. Rotated ring-attention steps may have different local Q and KV boundaries, so those steps now explicitly disable the fast path.
NOTE: This was found while strengthening the CP test with a nonuniform doutput as mentioned above

4. Make general THD path selection metadata-aware: Non-CP calls retain the fast path when Q and KV metadata match. Equal-length cross-attention with independent metadata uses the general path.

5. Strengthen bwd fused attn tests: 
    1. All bwd runner cases now use deterministic nonuniform doutput, including max-logit, score-mod, and distributed self/cross attention. The runner rejects an accidental missing seed. Same motivation as earlier.
    2. The cancellation-sensitive scalar objective comparison was removed; fwd outputs and bwd gradients continue to be compared element wise.
    3. Distributed BF16 dBias uses a reduction-aware default tolerance because its collective accumulation order differs from the unsharded reference.
 
6. Correct DP/FSDP test data placement (side effect of https://github.com/NVIDIA/TransformerEngine/pull/3272) :
    1. The eight-GPU dp2 × fsdp2 × tp2 test previously sharded activations over DP but replicated them over FSDP. TE then correctly reduced dBias across both axes, summing duplicate FSDP contributions.
    2. Batch-dependent test inputs and metadata are now sharded over the compound (dp, fsdp) axis. Related batch-divisibility checks use the combined data-parallel size.

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/jax/test_fused_attn.py [hunk 1]
```diff
@@ -1285,17 +1307,6 @@ def grad_func(
         if self.dropout_prob > 0.0:
             return
 
-        print_debug_tensor_stats(f"primitive_out", primitive_out)
-        print_debug_tensor_stats(f"reference_grad_valid", reference_out)
-        print_debug_tensor_stats(f"diff_grad", jnp.abs(primitive_out - reference_out))
-        assert_allclose(
```
Comment (jberchtold-nvidia) [inline tests/jax/test_fused_attn.py:1291]: Was this `assert_allclose` also part of the debugging code you're removing here? Thanks for removing the debug prints

File: transformer_engine/jax/attention.py [hunk 2]
```diff
@@ -777,11 +782,41 @@ def _segment_ids_pos_to_seqlens_offsets(
     # must route bottom-right masks to the slow path.
 
     # Fast path: O(T) per row.
-    if (
-        attn_mask_type.is_causal() and not attn_mask_type.is_bottom_right() and window_size is None
-    ) or (window_size == (-1, -1) and not attn_mask_type.is_bottom_right()):
-        return _segment_ids_pos_to_seqlens_offsets_fast_causal_path(
-            segment_ids_q, segment_ids_kv, segment_pos_q, segment_pos_kv, max_segments_per_seq
+    fast_path_requested = allow_fast_causal_path and (
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/attention.py:785]: What is the performance impact of not picking the fast path? 

If it is significant, it might be nice to log a warning with the reasoning the fast path was unable to be taken. This would give users insight if they see a sudden TE perf regression if they take a path where the fast path is not supported

File: transformer_engine/jax/attention.py [hunk 3]
```diff
@@ -847,7 +882,12 @@ def tree_unflatten(cls, aux_data, children):
         return cls(*children)
 
     def get_seqlens_and_offsets(
-        self, attn_mask_type, qkv_layout, window_size, max_segments_per_seq
+        self,
+        attn_mask_type,
+        qkv_layout,
+        window_size,
+        max_segments_per_seq,
+        allow_fast_causal_path=True,
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/attention.py:890]: When would a user want to intentionally set this to False? It seems internally we already automatically switch to a fallback if this `allow_fast_causal_path` is True but it isn't supported in that configuration.

If there isn't a valid use case for `allow_fast_causal_path=False` at the user level, maybe we keep this logic internal and don't expose this knob on this public API function.

If there is a valid use case where setting it to False is required/recommended, then let's expand the docstring to document that use case

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 4]
```diff
@@ -1063,6 +1068,20 @@ def shardy_sharding_rule(config, mesh, value_types, result_types):
...
+def _get_fused_attn_bwd_arg_shardings(arg_infos):
+    """Apply the common sharding constraints between backward operands."""
+    arg_shardings = [arg_i.sharding for arg_i in arg_infos]
+    # The fused backward kernel consumes output and doutput elementwise. Use the
+    # saved output rather than q: QKV-packed q has an additional packing axis.
+    output_idx = 7
+    doutput_idx = 8
+    arg_shardings[doutput_idx] = arg_shardings[output_idx]
+    # Each segment position tensor describes the tokens in its matching ID tensor.
+    arg_shardings[-1] = arg_shardings[-3]
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/cpp_extensions/attention.py:1080]: This is okay for now and pretty much the best we can do in the current sharding paradigm with automatic sharding propagation. But it could in theory trigger an AG and replication if one of the shardings was replicated, since it would copy that to the other arg's sharding. However, that is correct behavior since it ensures correct outputs and is mitigating the user-provided incorrect sharding

Good news, when we move to explicit sharding, the shardings are propagated at trace time so we can actually raise an error in cases like this and require the user to provide matching sharding. This will avoid cases like this where we have to do something safe but un-optimal (due to incorrect user-provided sharding)

Comment (jberchtold-nvidia) [reply transformer_engine/jax/cpp_extensions/attention.py:1080]: It's nice that you refactored this commonly re-used code into a helper so when we move to explicit sharding we can handle this error raising in one place

File: transformer_engine/jax/attention.py (hunk 2 above)
Comment (KshitijLakhani) [reply transformer_engine/jax/attention.py:785]: So in previous experiment that I have run, the general-path overhead was approximately 10-25% across some  cases IIRC so I would definitely say it is significant enough and that's why we maintain this separate path. 
I had thought about this a bit and here's my reasons for not throwing a warning:
-  The independent cross-attn and rotated THD CP are valid general-path cases : so there isn't really an alternative, so the warning could be informative for the case you mentioned (where they switch between input embeddings and see perf regression) but I do not think there's much they can do with it, tbh.
- IIUC, the decision is also inside lax.cond (run time), so a dynamic warning might require a host callback/debug print and could add overhead and repeated messages under JIT - or maybe I do not fully understand your suggestion ?

File: tests/jax/test_fused_attn.py (hunk 1 above)
Comment (KshitijLakhani) [reply tests/jax/test_fused_attn.py:1291]: It was not debug-only.
As part of testing, I figured that with signed random doutput, the BF16 reduction as part of code that I removed can be close to zero and sensitive to accumulation order. Since fwd outputs and dq/dk/dv/dbias are already checked element wise, I removed the redundant scalar comparison because it seemed like the better option rather than relaxing its tolerance (just to match the new seed).

File: transformer_engine/jax/cpp_extensions/attention.py (hunk 4 above)
Comment (KshitijLakhani) [reply transformer_engine/jax/cpp_extensions/attention.py:1080]: > This is okay for now and pretty much the best we can do in the current sharding paradigm with automatic sharding propagation. But it could in theory trigger an AG and replication if one of the shardings was replicated, since it would copy that to the other arg's sharding. However, that is correct behavior since it ensures correct outputs and is mitigating the user-provided incorrect sharding

Makes complete sense to me. Thanks for adding that note here. 

> Good news, when we move to explicit sharding, the shardings are propagated at trace time so we can actually raise an error in cases like this and require the user to provide matching sharding.

Definitely. That's a good point

File: transformer_engine/jax/attention.py (hunk 3 above)
Comment (KshitijLakhani) [reply transformer_engine/jax/attention.py]: You are absolutely right. Thanks for catching this. I had missed it :(
Pushed a commit for keeping the logic internal, as it should be.

File: transformer_engine/jax/attention.py (hunk 2 above)
Comment (jberchtold-nvidia) [reply transformer_engine/jax/attention.py:785]: Sounds good, resolved

File: tests/jax/test_fused_attn.py (hunk 1 above)
Comment (jberchtold-nvidia) [reply tests/jax/test_fused_attn.py:1291]: Seems reasonable, thanks

**Review by jberchtold-nvidia — APPROVED** [review]
